packages/react-query/src/__tests__/queryOptions.types.test.tsx

```diff
@@ -20,12 +25,156 @@ describe('queryOptions', () => {
+      return result
+    })
+  })
+  it('should work when passed to useQueries', () => {
+    doNotExecute(() => {
+      const options = queryOptions({
+        queryKey: ['key'],
+        queryFn: () => Promise.resolve(5),
+      })
+
+      const [{ data }] = useQueries({
+        queries: [options],
+      })
+
+      const result: Expect<Equal<typeof data, number | undefined>> = true
+      return result
+    })
+  })
+  it('should tag the queryKey with the result type of the QueryFn', () => {
+    doNotExecute(() => {
+      const { queryKey } = queryOptions({
+        queryKey: ['key'],
+        queryFn: () => Promise.resolve(5),
+      })
+
+      const result: Expect<
+        Equal<(typeof queryKey)[typeof dataTagSymbol], number>
+      > = true
+      return result
+    })
+
+    it('should tag the queryKey even if no promise is returned', () => {
+      doNotExecute(() => {
+        const { queryKey } = queryOptions({
+          queryKey: ['key'],
+          queryFn: () => 5,
+        })
+
+        const result: Expect<
+          Equal<(typeof queryKey)[typeof dataTagSymbol], number>
+        > = true
+        return result
+      })
+    })
+
+    it('should tag the queryKey with unknown if there is no queryFn', () => {
+      doNotExecute(() => {
+        const { queryKey } = queryOptions({
+          queryKey: ['key'],
+        })
+
+        const result: Expect<
+          Equal<(typeof queryKey)[typeof dataTagSymbol], unknown>
+        > = true
+        return result
+      })
+    })
+
+    it('should return the proper type when passed to getQueryData', () => {
+      doNotExecute(() => {
+        const { queryKey } = queryOptions({
+          queryKey: ['key'],
+          queryFn: () => Promise.resolve(5),
+        })
+
+        const queryClient = new QueryClient()
+        const data = queryClient.getQueryData(queryKey)
+
+        const result: Expect<Equal<typeof data, number | undefined>> = true
+        return result
+      })
+    })
+
+    it('should properly type updaterFn when passed to setQueryData', () => {
+      doNotExecute(() => {
+        const { queryKey } = queryOptions({
+          queryKey: ['key'],
+          queryFn: () => Promise.resolve(5),
+        })
+
+        const queryClient = new QueryClient()
+        const data = queryClient.setQueryData(queryKey, (prev) => {
+          const result: Expect<Equal<typeof prev, number | undefined>> = true
+          return result ? prev : 1
+        })
+
+        const result: Expect<Equal<typeof data, number>> = true
+        return result
+      })
+    })
+
+    it('should properly type value when passed to setQueryData', () => {
+      doNotExecute(() => {
+        const { queryKey } = queryOptions({
+          queryKey: ['key'],
+          queryFn: () => Promise.resolve(5),
+        })
+
+        const queryClient = new QueryClient()
+
+        // @ts-expect-error value should be a number
+        queryClient.setQueryData(queryKey, '5')
```

Comment:
@Andarist ideally, we would get an error here, because `queryKey` is tagged with `number`, but we pass in a `string` as value. However, I think it just takes the second overload and infers `string` because the first overload is not a match. Is there a way to make the first overload match and thus get an error here?

Note that it works fine if we use the functional updater form (test above this one), because `prev` will always be a `number` and we thus have to return a number as well.